File: backend/src/services/scoring.ts
import type {
  AccessibilityPoint,
  AccessibilityStatus,
  BonusEntry,
  Coordinates,
  EvidenceItem,
  PenaltyEntry,
  ProfilePreferences,
  RouteCandidate,
  RouteFactors,
} from "../types/index.js";
import {
  clamp,
  haversineDistance,
  pointToPolylineDistanceM,
  resamplePolyline,
} from "../utils/spatial.js";

/** Accessibility features further than this from the route line are not attached. */
export const EVIDENCE_RADIUS_M = 30;

/** A sampled route point is "known" when accessibility data exists within this radius. */
export const COVERAGE_RADIUS_M = 60;

/** Distance between samples used to estimate unknown coverage. */
export const SAMPLE_INTERVAL_M = 60;

/** Configurable, transparent scoring weights. */
export const WEIGHTS = {
  stairs: {
    wheelchair: 25,
    walker: 15,
    cane: 12,
    limited_mobility: 15,
    custom: 18,
  },
  steepSlope: 14,
  roughSurface: 10,
  obstacle: 22,
  unknownSegment: 6,
  unknownSegmentCap: 5,
  distancePerKm: 5,
  bonuses: {
    accessibleEntrance: 8,
    ramp: 6,
    elevator: 6,
    automaticDoor: 4,
    smoothSurface: 4,
    accessibleCrossing: 4,
    accessibleFeature: 2,
  },
} as const;

const TYPE_LABELS: Record<string, string> = {
  entrance: "Entrance",
  ramp: "Ramp",
  elevator: "Elevator",
  stairs: "Steps",
  crossing: "Crossing",
  automatic_door: "Automatic door",
  barrier: "Barrier",
  obstacle: "Obstacle",
  other: "Feature",
};

function typeLabel(type: string): string {
  return TYPE_LABELS[type] ?? "Feature";
}

function pointLabel(point: AccessibilityPoint): string {
  const base = typeLabel(point.type);
  return point.buildingName ? `${base} · ${point.buildingName}` : base;
}

/** Determine the accessibility status of a point for a given profile. */
export function statusForPoint(
  point: AccessibilityPoint,
  profile: ProfilePreferences,
): AccessibilityStatus {
  if (point.severity === "blocked") return "inaccessible";
  if (point.wheelchair && point.wheelchair !== "unknown") return point.wheelchair;
  if (point.type === "stairs" || point.stairs) {
    return profile.avoidStairs ? "inaccessible" : "unknown";
  }
  if (point.ramp || point.elevator || point.automaticDoor) return "accessible";
  if (point.surface === "rough") return "unknown";
  return "unknown";
}

export interface EvidenceResult {
  evidence: EvidenceItem[];
  factors: RouteFactors;
  unknownCoordinates: Coordinates[];
  totalSamples: number;
  knownSamples: number;
}

/**
 * Attach accessibility evidence to a route using proximity matching,
 * and sample the route to estimate how much of it is "known".
 */
export function buildEvidence(
  route: RouteCandidate,
  points: AccessibilityPoint[],
  profile: ProfilePreferences,
): EvidenceResult {
  const evidence: EvidenceItem[] = [];
  const factors: RouteFactors = {
    stairs: 0,
    ramps: 0,
    elevators: 0,
    crossings: 0,
    accessibleEntrances: 0,
    obstacles: 0,
    steepSlopes: 0,
    roughSurface: 0,
    unknownSections: 0,
    totalSamples: 0,
  };

  let minLat = Number.POSITIVE_INFINITY;
  let maxLat = Number.NEGATIVE_INFINITY;
  let minLon = Number.POSITIVE_INFINITY;
  let maxLon = Number.NEGATIVE_INFINITY;
  for (const c of route.geometry) {
    if (c.latitude < minLat) minLat = c.latitude;
    if (c.latitude > maxLat) maxLat = c.latitude;
    if (c.longitude < minLon) minLon = c.longitude;
    if (c.longitude > maxLon) maxLon = c.longitude;
  }
  const pad = EVIDENCE_RADIUS_M / 111000;
  const nearPoints = points.filter(
    (p) =>
      p.latitude >= minLat - pad &&
      p.latitude <= maxLat + pad &&
      p.longitude >= minLon - pad &&
      p.longitude <= maxLon + pad,
  );

  for (const point of nearPoints) {
    const distance = pointToPolylineDistanceM(
      { latitude: point.latitude, longitude: point.longitude },
      route.geometry,
    );
    if (distance > EVIDENCE_RADIUS_M) continue;

    const isStairs = point.type === "stairs" || point.stairs === true;
    const isRamp = point.type === "ramp" || point.ramp === true;
    const isElevator = point.type === "elevator" || point.elevator === true;
    const isCrossing = point.type === "crossing";
    const isEntrance = point.type === "entrance";
    const isObstacle =
      point.severity === "blocked" ||
      (point.type === "obstacle" && point.isTemporary === true);

    if (isStairs) factors.stairs += 1;
    if (isRamp) factors.ramps += 1;
    if (isElevator) factors.elevators += 1;
    if (isCrossing) factors.crossings += 1;
    if (isEntrance && point.wheelchair === "accessible") factors.accessibleEntrances += 1;
    if (isObstacle) factors.obstacles += 1;
    if (point.incline === "steep") factors.steepSlopes += 1;
    if (point.surface === "rough") factors.roughSurface += 1;

    const status = statusForPoint(point, profile);
    const severity = point.severity ?? "info";
    evidence.push({
      id: point.id,
      label: pointLabel(point),
      type: point.type,
      latitude: point.latitude,
      longitude: point.longitude,
      distanceMeters: Math.round(distance),
      sourceType: point.sourceType,
      status,
      severity,
      description: point.description,
      photoUrl: point.photoUrl,
      verified: point.sourceType === "community" ? Boolean(point.verifiedAt) : undefined,
    });
  }

  // Bucket nearby points into a uniform grid so coverage lookup is near O(1)
  // instead of scanning every point for every sample.
  const cellSizeDeg = COVERAGE_RADIUS_M / 111000;
  const grid = new Map<string, AccessibilityPoint[]>();
  for (const point of nearPoints) {
    const key = `${Math.floor(point.latitude / cellSizeDeg)},${Math.floor(
      point.longitude / cellSizeDeg,
    )}`;
    const bucket = grid.get(key);
    if (bucket) bucket.push(point);
    else grid.set(key, [point]);
  }

  const samples = resamplePolyline(route.geometry, SAMPLE_INTERVAL_M);
  let unknownCoordinates: Coordinates[] = [];
  let knownSamples = 0;
  const latRange = Math.ceil((COVERAGE_RADIUS_M / 111000) / cellSizeDeg);
  for (const sample of samples) {
    const cx = Math.floor(sample.latitude / cellSizeDeg);
    const cy = Math.floor(sample.longitude / cellSizeDeg);
    // Longitude degrees-per-metre shrink with cos(latitude), so widen the
    // search radius in the longitude axis to guarantee no in-range point is
    // missed (exact same result as the old all-points scan).
    const lonRange = Math.ceil(
      (COVERAGE_RADIUS_M / (111000 * Math.cos((sample.latitude * Math.PI) / 180))) /
        cellSizeDeg,
    );
    let known = false;
    for (let dx = -latRange; dx <= latRange && !known; dx++) {
      for (let dy = -lonRange; dy <= lonRange && !known; dy++) {
        const bucket = grid.get(`${cx + dx},${cy + dy}`);
        if (!bucket) continue;
        for (const point of bucket) {
          if (haversineDistance(sample, point) <= COVERAGE_RADIUS_M) {
            known = true;
            break;
          }
        }
      }
    }
    if (known) knownSamples += 1;
    else unknownCoordinates.push(sample);
  }
  factors.unknownSections = samples.length - knownSamples;
  factors.totalSamples = samples.length;

  return { evidence, factors, unknownCoordinates, totalSamples: samples.length, knownSamples };
}

export interface ScoreResult {
  score: number;
  penalties: PenaltyEntry[];
  bonuses: BonusEntry[];
}

/** Compute the explainable 0-100 accessibility score for a route. */
export function scoreRoute(
  route: RouteCandidate,
  evidenceResult: EvidenceResult,
  profile: ProfilePreferences,
): ScoreResult {
  const { evidence, factors } = evidenceResult;
  const penalties: PenaltyEntry[] = [];
  const bonuses: BonusEntry[] = [];

  const stairsWeight =
    profile.mobilityProfile === "custom" && profile.avoidStairs
      ? WEIGHTS.stairs.custom
      : WEIGHTS.stairs[profile.mobilityProfile] ?? WEIGHTS.stairs.walker;

  // Penalties are aggregated by category and capped so a handful of bad
  // features can't drive every downtown route straight to 0 (and vice versa).
  const stairItems = evidence.filter((item) => item.type === "stairs");
  if (stairItems.length > 0) {
    penalties.push({
      label: `${stairItems.length} step section${stairItems.length > 1 ? "s" : ""}`,
      points: stairsWeight * Math.min(stairItems.length, 2),
      severity: profile.mobilityProfile === "wheelchair" ? "critical" : "warning",
      detail: stairItems[0]?.description,
    });
  }

  const blockedItems = evidence.filter((item) => item.severity === "blocked");
  if (blockedItems.length > 0) {
    penalties.push({
      label: `${blockedItems.length} blocked feature${blockedItems.length > 1 ? "s" : ""}`,
      points: WEIGHTS.obstacle * Math.min(blockedItems.length, 2),
      severity: "critical",
      detail: blockedItems[0]?.description,
    });
  }

  const obstacleWarnings = evidence.filter(
    (item) => item.type === "obstacle" && item.severity === "warning",
  );
  if (obstacleWarnings.length > 0) {
    penalties.push({
      label: `${obstacleWarnings.length} obstacle${obstacleWarnings.length > 1 ? "s" : ""} to work around`,
      points: WEIGHTS.roughSurface * Math.min(obstacleWarnings.length, 2),
      severity: "warning",
      detail: obstacleWarnings[0]?.description,
    });
  }

  if (factors.steepSlopes > 0) {
    penalties.push({
      label: `${factors.steepSlopes} steep slope${factors.steepSlopes > 1 ? "s" : ""}`,
      points: WEIGHTS.steepSlope * Math.min(factors.steepSlopes, 2),
      severity: "warning",
    });
  }
  if (factors.roughSurface > 0) {
    penalties.push({
      label: `${factors.roughSurface} rough surface section${factors.roughSurface > 1 ? "s" : ""}`,
      points: WEIGHTS.roughSurface * Math.min(factors.roughSurface, 2),
      severity: "warning",
    });
  }
  if (factors.unknownSections > 0) {
    penalties.push({
      label: `${factors.unknownSections} route section${factors.unknownSections > 1 ? "s" : ""} without accessibility data`,
      points:
        WEIGHTS.unknownSegment * Math.min(factors.unknownSections, WEIGHTS.unknownSegmentCap),
      severity: "info",
      detail: "Unknown is not 'inaccessible' - it just means we lack data here.",
    });
  }

  const rampBonus = profile.preferRamps ? WEIGHTS.bonuses.ramp : 3;
  const elevatorBonus = profile.preferElevators ? WEIGHTS.bonuses.elevator : 3;

  const countOf = (pred: (item: EvidenceItem) => boolean) => evidence.filter(pred).length;
  const entranceCount = countOf((i) => i.type === "entrance" && i.status === "accessible");
  const rampCount = countOf((i) => i.type === "ramp");
  const elevatorCount = countOf((i) => i.type === "elevator" && i.severity !== "blocked");
  const doorCount = countOf((i) => i.type === "automatic_door");
  const crossingCount = countOf((i) => i.type === "crossing" && i.status === "accessible");

  if (entranceCount > 0) {
    bonuses.push({
      label: `${entranceCount} accessible entrance${entranceCount > 1 ? "s" : ""}`,
      points: WEIGHTS.bonuses.accessibleEntrance * Math.min(entranceCount, 2),
    });
  }
  if (rampCount > 0) {
    bonuses.push({
      label: `${rampCount} ramp${rampCount > 1 ? "s" : ""}`,
      points: rampBonus * Math.min(rampCount, 2),
    });
  }
  if (elevatorCount > 0) {
    bonuses.push({
      label: `${elevatorCount} elevator${elevatorCount > 1 ? "s" : ""}`,
      points: elevatorBonus * Math.min(elevatorCount, 2),
    });
  }
  if (doorCount > 0) {
    bonuses.push({
      label: `${doorCount} automatic door${doorCount > 1 ? "s" : ""}`,
      points: WEIGHTS.bonuses.automaticDoor * Math.min(doorCount, 2),
    });
  }
  if (crossingCount > 0) {
    bonuses.push({
      label: `${crossingCount} accessible crossing${crossingCount > 1 ? "s" : ""}`,
      points: WEIGHTS.bonuses.accessibleCrossing * Math.min(crossingCount, 3),
    });
  }

  const km = route.distanceMeters / 1000;
  const distancePenalty = Math.round(WEIGHTS.distancePerKm * km * 10) / 10;
  if (distancePenalty > 0) {
    penalties.push({
      label: `Distance (${route.distanceMeters} m)`,
      points: distancePenalty,
      severity: "info",
    });
  }

  const sumPenalties = penalties.reduce((sum, p) => sum + p.points, 0);
  const sumBonuses = bonuses.reduce((sum, b) => sum + b.points, 0);
  const score = clamp(
    Math.round(100 - sumPenalties + Math.min(sumBonuses, 30)),
    0,
    100,
  );

  return { score, penalties, bonuses };
}

File: backend/app/services/scoring.py
"""Accessibility evidence attachment and explainable route scoring.

Port of the Node scoring service. Weights are centralised and configurable,
missing data is treated as unknown (never as a sign of accessibility)."""

import math

from ..schemas import (
    AccessibilityPoint,
    AccessibilityStatus,
    BonusEntry,
    Coordinates,
    EvidenceItem,
    PenaltyEntry,
    ProfilePreferences,
    RouteCandidate,
    RouteFactors,
)
from ..utils.spatial import (
    clamp,
    haversine_distance,
    point_to_polyline_distance_m,
    resample_polyline,
)

EVIDENCE_RADIUS_M = 45
COVERAGE_RADIUS_M = 60
SAMPLE_INTERVAL_M = 60

WEIGHTS = {
    "stairs": {
        "wheelchair": 25,
        "walker": 15,
        "cane": 12,
        "limited_mobility": 15,
        "custom": 18,
    },
    "steepSlope": 14,
    "roughSurface": 10,
    "obstacle": 22,
    "unknownSegment": 6,
    "unknownSegmentCap": 5,
    "distancePerKm": 5,
    "bonuses": {
        "accessibleEntrance": 8,
        "ramp": 6,
        "elevator": 6,
        "automaticDoor": 4,
        "smoothSurface": 4,
        "accessibleCrossing": 4,
        "accessibleFeature": 2,
    },
}

TYPE_LABELS = {
    "entrance": "Entrance",
    "ramp": "Ramp",
    "elevator": "Elevator",
    "stairs": "Steps",
    "crossing": "Crossing",
    "automatic_door": "Automatic door",
    "barrier": "Barrier",
    "obstacle": "Obstacle",
    "other": "Feature",
}


def type_label(type_: str) -> str:
    return TYPE_LABELS.get(type_, "Feature")


def point_label(point: AccessibilityPoint) -> str:
    base = type_label(point.type)
    return f"{base} · {point.buildingName}" if point.buildingName else base


def status_for_point(point: AccessibilityPoint, profile: ProfilePreferences) -> AccessibilityStatus:
    if point.severity == "blocked":
        return "inaccessible"
    if point.wheelchair and point.wheelchair != "unknown":
        return point.wheelchair
    if point.type == "stairs" or point.stairs:
        return "inaccessible" if profile.avoidStairs else "unknown"
    if point.ramp or point.elevator or point.automaticDoor:
        return "accessible"
    if point.surface == "rough":
        return "unknown"
    return "unknown"


class EvidenceResult:
    def __init__(self) -> None:
        self.evidence: list[EvidenceItem] = []
        self.factors: RouteFactors = RouteFactors()
        self.unknown_coordinates: list[Coordinates] = []
        self.total_samples = 0
        self.known_samples = 0


def build_evidence(
    route: RouteCandidate,
    points: list[AccessibilityPoint],
    profile: ProfilePreferences,
) -> EvidenceResult:
    result = EvidenceResult()

    min_lat = float("inf")
    max_lat = float("-inf")
    min_lon = float("inf")
    max_lon = float("-inf")
    for c in route.geometry:
        min_lat = min(min_lat, c.latitude)
        max_lat = max(max_lat, c.latitude)
        min_lon = min(min_lon, c.longitude)
        max_lon = max(max_lon, c.longitude)

    pad = EVIDENCE_RADIUS_M / 111000
    near_points = [
        p
        for p in points
        if min_lat - pad <= p.latitude <= max_lat + pad
        and min_lon - pad <= p.longitude <= max_lon + pad
    ]

    factors = result.factors
    for point in near_points:
        distance = point_to_polyline_distance_m(
            Coordinates(latitude=point.latitude, longitude=point.longitude),
            route.geometry,
        )
        if distance > EVIDENCE_RADIUS_M:
            continue

        is_stairs = point.type == "stairs" or point.stairs is True
        is_ramp = point.type == "ramp" or point.ramp is True
        is_elevator = point.type == "elevator" or point.elevator is True
        is_crossing = point.type == "crossing"
        is_entrance = point.type == "entrance"
        is_obstacle = (
            point.severity == "blocked"
            or (point.type == "obstacle" and point.isTemporary is True)
        )

        if is_stairs:
            factors.stairs += 1
        if is_ramp:
            factors.ramps += 1
        if is_elevator:
            factors.elevators += 1
        if is_crossing:
            factors.crossings += 1
        if is_entrance and point.wheelchair == "accessible":
            factors.accessibleEntrances += 1
        if is_obstacle:
            factors.obstacles += 1
        if point.incline == "steep":
            factors.steepSlopes += 1
        if point.surface == "rough":
            factors.roughSurface += 1

        status = status_for_point(point, profile)
        severity = point.severity or "info"
        verified = bool(point.verifiedAt) if point.sourceType == "community" else None
        result.evidence.append(
            EvidenceItem(
                id=point.id,
                label=point_label(point),
                type=point.type,
                latitude=point.latitude,
                longitude=point.longitude,
                distanceMeters=round(distance),
                sourceType=point.sourceType,
                status=status,
                severity=severity,
                description=point.description,
                photoUrl=point.photoUrl,
                verified=verified,
            )
        )

    cell_size_deg = COVERAGE_RADIUS_M / 111000
    grid: dict[tuple[int, int], list[AccessibilityPoint]] = {}
    for point in near_points:
        key = (math.floor(point.latitude / cell_size_deg), math.floor(point.longitude / cell_size_deg))
        grid.setdefault(key, []).append(point)

    samples = resample_polyline(route.geometry, SAMPLE_INTERVAL_M)
    unknown_coordinates: list[Coordinates] = []
    known_samples = 0
    lat_range = math.ceil((COVERAGE_RADIUS_M / 111000) / cell_size_deg)
    for sample in samples:
        cx = math.floor(sample.latitude / cell_size_deg)
        cy = math.floor(sample.longitude / cell_size_deg)
        lon_range = math.ceil(
            (COVERAGE_RADIUS_M / (111000 * math.cos(sample.latitude * math.pi / 180)))
            / cell_size_deg
        )
        known = False
        for dx in range(-lat_range, lat_range + 1):
            if known:
                break
            for dy in range(-lon_range, lon_range + 1):
                if known:
                    break
                bucket = grid.get((cx + dx, cy + dy))
                if not bucket:
                    continue
                for point in bucket:
                    if haversine_distance(sample, point) <= COVERAGE_RADIUS_M:
                        known = True
                        break
        if known:
            known_samples += 1
        else:
            unknown_coordinates.append(sample)

    factors.unknownSections = len(samples) - known_samples
    factors.totalSamples = len(samples)

    result.unknown_coordinates = unknown_coordinates
    result.total_samples = len(samples)
    result.known_samples = known_samples
    return result


class ScoreResult:
    def __init__(self, score: int, penalties: list[PenaltyEntry], bonuses: list[BonusEntry]):
        self.score = score
        self.penalties = penalties
        self.bonuses = bonuses


def score_route(
    route: RouteCandidate,
    evidence_result: EvidenceResult,
    profile: ProfilePreferences,
) -> ScoreResult:
    evidence = evidence_result.evidence
    factors = evidence_result.factors
    penalties: list[PenaltyEntry] = []
    bonuses: list[BonusEntry] = []

    base_stairs_weight = (
        WEIGHTS["stairs"]["custom"]
        if profile.mobilityProfile == "custom" and profile.avoidStairs
        else WEIGHTS["stairs"].get(profile.mobilityProfile, WEIGHTS["stairs"]["walker"])
    )
    stairs_weight = (
        base_stairs_weight if profile.avoidStairs else max(3, round(base_stairs_weight * 0.4))
    )

    stair_items = [item for item in evidence if item.type == "stairs"]
    if stair_items:
        label = (
            f"{len(stair_items)} step section"
            if len(stair_items) == 1
            else f"{len(stair_items)} step sections"
        )
        penalties.append(
            PenaltyEntry(
                label=label,
                points=stairs_weight * min(len(stair_items), 2),
                severity="critical" if profile.mobilityProfile == "wheelchair" else "warning",
                detail=stair_items[0].description,
            )
        )

    blocked_items = [item for item in evidence if item.severity == "blocked"]
    if blocked_items:
        label = (
            f"{len(blocked_items)} blocked feature"
            if len(blocked_items) == 1
            else f"{len(blocked_items)} blocked features"
        )
        penalties.append(
            PenaltyEntry(
                label=label,
                points=WEIGHTS["obstacle"] * min(len(blocked_items), 2),
                severity="critical",
                detail=blocked_items[0].description,
            )
        )

    obstacle_warnings = [
        item for item in evidence if item.type == "obstacle" and item.severity == "warning"
    ]
    if obstacle_warnings:
        label = (
            f"{len(obstacle_warnings)} obstacle to work around"
            if len(obstacle_warnings) == 1
            else f"{len(obstacle_warnings)} obstacles to work around"
        )
        penalties.append(
            PenaltyEntry(
                label=label,
                points=WEIGHTS["roughSurface"] * min(len(obstacle_warnings), 2),
                severity="warning",
                detail=obstacle_warnings[0].description,
            )
        )

    slope_weight = {
        "flat": WEIGHTS["steepSlope"],
        "moderate": WEIGHTS["steepSlope"],
        "steep": 4,
        "any": 0,
    }.get(profile.maxSlope, WEIGHTS["steepSlope"])
    if factors.steepSlopes > 0 and slope_weight > 0:
        label = (
            f"{factors.steepSlopes} steep slope"
            if factors.steepSlopes == 1
            else f"{factors.steepSlopes} steep slopes"
        )
        penalties.append(
            PenaltyEntry(
                label=label,
                points=slope_weight * min(factors.steepSlopes, 2),
                severity="warning",
            )
        )
    rough_weight = WEIGHTS["roughSurface"] if profile.preferSmoothSurface else 3
    if factors.roughSurface > 0 and rough_weight > 0:
        label = (
            f"{factors.roughSurface} rough surface section"
            if factors.roughSurface == 1
            else f"{factors.roughSurface} rough surface sections"
        )
        penalties.append(
            PenaltyEntry(
                label=label,
                points=rough_weight * min(factors.roughSurface, 2),
                severity="warning",
            )
        )
    if factors.unknownSections > 0:
        label = (
            f"{factors.unknownSections} route section without accessibility data"
            if factors.unknownSections == 1
            else f"{factors.unknownSections} route sections without accessibility data"
        )
        penalties.append(
            PenaltyEntry(
                label=label,
                points=WEIGHTS["unknownSegment"]
                * min(factors.unknownSections, WEIGHTS["unknownSegmentCap"]),
                severity="info",
                detail="Unknown is not 'inaccessible' - it just means we lack data here.",
            )
        )

    ramp_bonus = WEIGHTS["bonuses"]["ramp"] if profile.preferRamps else 3
    elevator_bonus = WEIGHTS["bonuses"]["elevator"] if profile.preferElevators else 3

    def count_of(pred) -> int:
        return sum(1 for item in evidence if pred(item))

    entrance_count = count_of(lambda i: i.type == "entrance" and i.status == "accessible")
    ramp_count = count_of(lambda i: i.type == "ramp")
    elevator_count = count_of(lambda i: i.type == "elevator" and i.severity != "blocked")
    door_count = count_of(lambda i: i.type == "automatic_door")
    crossing_count = count_of(lambda i: i.type == "crossing" and i.status == "accessible")

    if entrance_count > 0:
        bonuses.append(
            BonusEntry(
                label=(
                    f"{entrance_count} accessible entrance"
                    if entrance_count == 1
                    else f"{entrance_count} accessible entrances"
                ),
                points=WEIGHTS["bonuses"]["accessibleEntrance"] * min(entrance_count, 2),
            )
        )
    if ramp_count > 0:
        bonuses.append(
            BonusEntry(
                label=f"{ramp_count} ramp" if ramp_count == 1 else f"{ramp_count} ramps",
                points=ramp_bonus * min(ramp_count, 2),
            )
        )
    if elevator_count > 0:
        bonuses.append(
            BonusEntry(
                label=f"{elevator_count} elevator"
                if elevator_count == 1
                else f"{elevator_count} elevators",
                points=elevator_bonus * min(elevator_count, 2),
            )
        )
    if door_count > 0:
        bonuses.append(
            BonusEntry(
                label=f"{door_count} automatic door"
                if door_count == 1
                else f"{door_count} automatic doors",
                points=WEIGHTS["bonuses"]["automaticDoor"] * min(door_count, 2),
            )
        )
    if crossing_count > 0:
        bonuses.append(
            BonusEntry(
                label=f"{crossing_count} accessible crossing"
                if crossing_count == 1
                else f"{crossing_count} accessible crossings",
                points=WEIGHTS["bonuses"]["accessibleCrossing"] * min(crossing_count, 3),
            )
        )

    km = route.distanceMeters / 1000
    distance_penalty = round(WEIGHTS["distancePerKm"] * km * 10) / 10
    if distance_penalty > 0:
        penalties.append(
            PenaltyEntry(
                label=f"Distance ({route.distanceMeters} m)",
                points=distance_penalty,
                severity="info",
            )
        )

    sum_penalties = sum(p.points for p in penalties)
    sum_bonuses = sum(b.points for b in bonuses)
    score = clamp(round(100 - sum_penalties + min(sum_bonuses, 30)), 0, 100)

    return ScoreResult(score=int(score), penalties=penalties, bonuses=bonuses)
